### PR TITLE
Diff updated indexes using old view model

### PR DESCRIPTION
**Pull request by Goos**
Changes so that the HUBViewModelDiff calculates updated indices using the previous data model rather than the new one.

@spotify/objc-dev (@cerihughes, @JohnSundell @rastersize)

### CODE REVIEW

File: tests/HUBViewModelDiffTests.m [hunk 1]
```diff
@@ -184,10 +184,68 @@ - (void)testComplexChangeSet
...
+    XCTAssert([diff.reloadedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:8 inSection:0]]);
     XCTAssert(diff.reloadedBodyComponentIndexPaths.count == 2);
     XCTAssert(diff.insertedBodyComponentIndexPaths.count == 2);
     XCTAssert(diff.deletedBodyComponentIndexPaths.count == 2);
 }
 
+- (void)testInsertionOfComponentModelAtStartWithDataChanges
+{
+    NSArray<id<HUBComponentModel>> *firstComponents = @[
+                                                        [self createComponentModelWithIdentifier:@"component-1" customData:nil],
```
Comment (JohnSundell) [inline tests/HUBViewModelDiffTests.m]: Looks like this code formatting is a bit off 🎨

File: tests/HUBViewModelDiffTests.m [hunk 2]
```diff
@@ -184,10 +184,68 @@ - (void)testComplexChangeSet
     XCTAssert([diff.insertedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:2 inSection:0]]);
     XCTAssert([diff.insertedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:9 inSection:0]]);
     XCTAssert([diff.reloadedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:1 inSection:0]]);
-    XCTAssert([diff.reloadedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:7 inSection:0]]);
+    XCTAssert([diff.reloadedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:8 inSection:0]]);
```
Comment (JohnSundell) [inline tests/HUBViewModelDiffTests.m]: This seems a bit dangerous, or was the test incorrect?

File: tests/HUBViewModelDiffTests.m (hunk 1 above)
Comment (JohnSundell) [inline tests/HUBViewModelDiffTests.m]: (also the same for the other array literals)

File: tests/HUBViewModelDiffTests.m [hunk 3]
```diff
@@ -184,10 +184,68 @@ - (void)testComplexChangeSet
...
+    XCTAssert(diff.reloadedBodyComponentIndexPaths.count == 3);
+    XCTAssert(diff.insertedBodyComponentIndexPaths.count == 1);
+    XCTAssert(diff.deletedBodyComponentIndexPaths.count == 0);
+    XCTAssert([diff.insertedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:0 inSection:0]]);
+    XCTAssert([diff.reloadedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:0 inSection:0]]);
+    XCTAssert([diff.reloadedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:1 inSection:0]]);
+    XCTAssert([diff.reloadedBodyComponentIndexPaths containsObject:[NSIndexPath indexPathForItem:2 inSection:0]]);
+}
+
+- (void)testInsertionOfComponentModelsAtStartWithDataChanges
```
Comment (JohnSundell) [inline tests/HUBViewModelDiffTests.m]: Suggestion: I had to spend quite a few minutes to figure out what the difference was between this test and the one above. Renaming it to `testInsertionOfMultipleComponentModelsAtStartWithDataChanges` would make it a bit more explicit (IMO).

Comment (cerihughes) [inline tests/HUBViewModelDiffTests.m]: Sorry...  they were tests I put together quickly.

File: tests/HUBViewModelDiffTests.m (hunk 1 above)
Comment (cerihughes) [inline tests/HUBViewModelDiffTests.m]: Isn't that the default Xcode formatter settings? (it is on my setup at least)

**Comment by cerihughes** [comment]
@JohnSundell This should be used in preference to what I was working on (but we've pulled in the tests I wrote that reproduced the problem). Mine was a quick patch on top and needed an extra iteration, but this fixes the problem in the same number of iterations.

File: tests/HUBViewModelDiffTests.m (hunk 2 above)
Comment (cerihughes) [inline tests/HUBViewModelDiffTests.m]: The implementation we had made some incorrect assumptions about the index paths you need to pass to `UICollectionView performBatchUpdates:completion:` so would sometimes generate incorrect index paths that caused the crash.

The new code generates correct index paths, so this test has to be updated to reflect the correct logic.

**Comment by cerihughes** [comment]
Looks great 🎉

File: tests/HUBViewModelDiffTests.m (hunk 1 above)
Comment (JohnSundell) [inline tests/HUBViewModelDiffTests.m]: Yeah it is, but our allmighty [code style](https://github.com/spotify/ios-style) says differently 😄

**Comment by JohnSundell** [comment]
@cerihughes Ah got it, awesome! 👍

**Comment by JohnSundell** [comment]
@Goos @cerihughes Really awesome work guys! ⭐️ ⭐️ ⭐️ ⭐️ ⭐️

File: tests/HUBViewModelDiffTests.m (hunk 2 above)
Comment (JohnSundell) [inline tests/HUBViewModelDiffTests.m]: Ah makes sense 👍

Comment (Goos) [inline tests/HUBViewModelDiffTests.m]: This is due to the change – the updated item was at index 8 prior to the update :smile:

**Comment by Goos** [comment]
@goos, @cerihughes Updated the tests and fixed the style issues 😄

File: tests/HUBCollectionViewLayoutTests.m [hunk 4]
```diff
@@ -339,17 +339,17 @@ - (void)testProposedContentOffsetAfterRecomputing
...
+    [addedIndices enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
+        [self addBodyComponentWithIdentifier:self.fullWidthComponentIdentifier preferredIndex:idx];
+        if (idx < currentIndex) {
+            insertionHeight += self.fullWidthComponent.preferredViewSize.height;
         }
     }];
 
-    [addedIndices enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
-        [self addBodyComponentWithIdentifier:self.fullWidthComponentIdentifier preferredIndex:idx];
+    [removedIndices enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
```
Comment (JohnSundell) [inline tests/HUBCollectionViewLayoutTests.m]: Could you just give some quick context as to why these tests had to be changed? 😄

Comment (Goos) [inline tests/HUBCollectionViewLayoutTests.m]: It was actually a problem with the test that I hadn't seen before :smile: It wasn't adding & removing the intended components

Comment (JohnSundell) [inline tests/HUBCollectionViewLayoutTests.m]: Ah alright - thanks 🌮

**Pull request merged**